### PR TITLE
[FLINK-31114][runtime] Set parallelism of job vertices in forward group at compilation phase 

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change
To fix FLINK-31114.

## Verifying this change
`AdaptiveBatchSchedulerITCase#testDifferentConsumerParallelism`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchScheduler.java [hunk 1]
```diff
@@ -465,13 +465,7 @@ public static VertexParallelismStore computeVertexParallelismStoreForDynamicGrap
...
         return computeVertexParallelismStore(
                 vertices,
-                v -> {
-                    if (v.getParallelism() > 0) {
-                        return getDefaultMaxParallelism(v);
-                    } else {
-                        return defaultMaxParallelism;
-                    }
-                },
+                v -> Math.max(defaultMaxParallelism, v.getParallelism()),
```
Comment (JunRuiLee) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchScheduler.java:468]: Why do we need this change? Is this change related to bug fixes?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResult.java [hunk 2]
```diff
@@ -175,35 +175,28 @@ int getNumParallelProducers() {
...
-        List<JobEdge> consumers = intermediateDataSet.getConsumers();
-        checkState(!consumers.isEmpty());
-
-        InternalExecutionGraphAccessor graph = getProducer().getGraph();
-        int consumersParallelism =
-                graph.getJobVertex(consumers.get(0).getTarget().getID()).getParallelism();
-        if (consumers.size() == 1) {
-            return consumersParallelism;
+    int getMaxConsumerParallelism() {
+        if (getProducer().getGraph().isDynamic()) {
```
Comment (JunRuiLee) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResult.java:179]: It seems that the dynamic graph does not require the parallelism of its consumers to be consistent for partition reuse, but the non-dynamic graph is different. What caused this difference? In other words, can we reset the operator's parallelism at the time of jobGraph compilation and before setVertexConfig instead of in the scheduler like FLINK-30684.

**Review by JunRuiLee — COMMENTED** [review]
@wanglijie95 Thanks for creating this pr, I have a few comments. Please take a look.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 3]
```diff
@@ -1060,14 +1111,97 @@ private void setVertexConfig(
         vertexConfigs.put(vertexID, config);
     }
 
+    private void setChainedOutputsConfig(
+            Integer vertexId, StreamConfig config, List<StreamEdge> chainableOutputs) {
+        // iterate edges, find sideOutput edges create and save serializers for each outputTag type
```
Comment (JunRuiLee) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1116]: VertexId is a redundant parameter

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 4]
```diff
@@ -1060,14 +1111,97 @@ private void setVertexConfig(
...
+        jobVertices
+                .keySet()
+                .forEach(
+                        startNodeId -> {
+                            setVertexNonChainedOutputsConfig(
+                                    startNodeId,
+                                    vertexConfigs.get(startNodeId),
+                                    chainInfos.get(startNodeId).getTransitiveOutEdges(),
+                                    opIntermediateOutputs);
+                        });
```
Comment (JunRuiLee) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1194]: {} can be removed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroupComputeUtilTest.java [hunk 5]
```diff
@@ -176,58 +178,14 @@ void testOneInputSplitsIntoTwo() throws Exception {
...
-        checkGroupSize(groups, 2, 2, 2);
-        assertThat(v1.getParallelism()).isEqualTo(8);
-        assertThat(v2.getParallelism()).isEqualTo(8);
-        assertThat(v3.getParallelism()).isEqualTo(-1);
-        assertThat(v4.getParallelism()).isEqualTo(-1);
-    }
-
-    private static Set<ForwardGroup> computeForwardGroupsAndSetVertexParallelismsIfNecessary(
-            JobVertex... vertices) throws Exception {
+    private static Set<ForwardGroup> computeForwardGroups(JobVertex... vertices) throws Exception {
```
Comment (JunRuiLee) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroupComputeUtilTest.java:181]: Don't need throws exception

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroupComputeUtilTest.java [hunk 6]
```diff
@@ -36,7 +36,9 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-/** Unit tests for {@link ForwardGroupComputeUtil}. */
+/**
+ * Unit tests for {@link org.apache.flink.runtime.jobgraph.forwardgroup.ForwardGroupComputeUtil}.
+ */
 class ForwardGroupComputeUtilTest {
     @RegisterExtension
     static final TestExecutorExtension<ScheduledExecutorService> EXECUTOR_RESOURCE =
```
Comment (JunRuiLee) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroupComputeUtilTest.java:44]: Can be removed because the only use is removed in FLINK-30942.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 7]
```diff
@@ -928,14 +1011,8 @@ private StreamConfig createJobVertex(Integer streamNodeId, OperatorChainInfo cha
         return new StreamConfig(jobVertex.getConfiguration());
```
Comment (JunRuiLee) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1011]: Can we move the above logic of reset parallelism to setVertexParallelismForDynamicGraphIfNecessary?

**Review by JunRuiLee — COMMENTED** [review]
@wanglijie95 Thanks for updating! I've left some comments. PTAL

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 8]
```diff
@@ -928,14 +1011,8 @@ private StreamConfig createJobVertex(Integer streamNodeId, OperatorChainInfo cha
...
-    private void setVertexConfig(
-            Integer vertexID,
-            StreamConfig config,
-            List<StreamEdge> chainableOutputs,
-            List<StreamEdge> nonChainableOutputs,
-            Map<Integer, ChainedSourceInfo> chainedSources) {
-
-        tryConvertPartitionerForDynamicGraph(chainableOutputs, nonChainableOutputs);
+    private void setOperatorConfig(
+            Integer vertexID, StreamConfig config, Map<Integer, ChainedSourceInfo> chainedSources) {
```
Comment (zhuzhurk) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1015]: vertexID -> vertexId

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 9]
```diff
@@ -1955,6 +2091,14 @@ private OperatorID addNodeToChain(int currentNodeId, String operatorName) {
             return new OperatorID(primaryHashBytes);
         }
 
+        private void setTransitiveOutEdges(final List<StreamEdge> transitiveOutEdges) {
+            this.transitiveOutEdges.addAll(transitiveOutEdges);
+        }
+
+        public List<StreamEdge> getTransitiveOutEdges() {
```
Comment (zhuzhurk) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:2098]:
```suggestion
        private List<StreamEdge> getTransitiveOutEdges() {
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 10]
```diff
@@ -568,12 +592,7 @@ && isChainableInput(sourceOutEdge, streamGraph)) {
                     final StreamConfig.SourceInputConfig inputConfig =
                             new StreamConfig.SourceInputConfig(sourceOutEdge);
                     final StreamConfig operatorConfig = new StreamConfig(new Configuration());
-                    setVertexConfig(
-                            sourceNodeId,
-                            operatorConfig,
-                            Collections.emptyList(),
-                            Collections.emptyList(),
-                            Collections.emptyMap());
+                    setOperatorConfig(sourceNodeId, operatorConfig, Collections.emptyMap());
```
Comment (zhuzhurk) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:595]: IIUC, `setChainedOutputs(emptyList)` would be called here previously, but is no longer called after this change?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 11]
```diff
@@ -758,6 +771,76 @@ private List<StreamEdge> createChain(
         }
     }
 
+    private void setVertexParallelismsForDynamicGraphIfNecessary() {
```
Comment (zhuzhurk) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:774]: Better to add some comments to explain what this method is for, e.g. ensures the parallelism and maxParallelism of vertices in the same forward group to be the same; set the parallelism at early stage if possible, to avoid invalid partition reuse.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 4 above)
Comment (wanglijie95) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1194]: The {} is needed because the `setVertexNonChainedOutputsConfig` has no return value.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 7 above)
Comment (wanglijie95) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1011]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroup.java [hunk 12]
```diff
@@ -53,9 +55,22 @@ public ForwardGroup(final Set<JobVertex> jobVertices) {
...
+        Set<Integer> configuredMaxParallelisms =
+                jobVertices.stream()
+                        .map(JobVertex::getMaxParallelism)
+                        .filter(val -> val > 0)
+                        .collect(Collectors.toSet());
+
+        if (!configuredMaxParallelisms.isEmpty()) {
+            this.maxParallelism = Collections.min(configuredMaxParallelisms);
+            checkState(
+                    maxParallelism >= parallelism,
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/forwardgroup/ForwardGroup.java:72]: I will change the check to `checkState(parallelism == -1 || maxParallelism >= parallelism)`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 10 above)
Comment (wanglijie95) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:595]: Yes. I will modify it to the same logic as before

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 4 above)
Comment (wanglijie95) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1194]: I'm wrong. It's can be removed :)

**Comment by wanglijie95** [comment]
Thanks for review @JunRuiLee @zhuzhurk . I 've addressed all comments, please take a look.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Comment by JunRuiLee** [comment]
LGTM.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 13]
```diff
@@ -205,7 +216,7 @@ private StreamingJobGraphGenerator(
         this.defaultStreamGraphHasher = new StreamGraphHasherV2();
         this.legacyStreamGraphHashers = Arrays.asList(new StreamGraphUserHashHasher());
 
-        this.jobVertices = new HashMap<>();
+        this.jobVertices = new LinkedHashMap<>();
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:219]: Yes, it's needed, because we hope to set vertex/operator parallelisms/outputs in the same order as vertices' creation. I
I think the changes of order will not have any negative impacts ?
